### PR TITLE
Improved Linter

### PR DESCRIPTION
**Pull request by bkallaher**
The linter is now more fully featured for what we want without cluttering the CMakeLists.txt file.

### CODE REVIEW

File: tools/linters/cpplint_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+import os.path
+import re
+
+
+def patch(original_module):
+    """ Decorator to easily allow wrapping/overriding of the Check* functions in cpplint. Should
+        decorate a function which matches the signature of the function it replaces expect with
+        the addition of a fn parameter, which is a pass-through of the replaced function, in case
+        the replacement would like call through to the original functionality. """
```
Comment (skallaher) [inline tools/linters/cpplint_wrapper.py]: Is `expect` meant to be "except"? Also "would like call"?

File: tools/linters/cpplint_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        original_fn = getattr(original_module, override_fn.__name__)
+        setattr(original_module, override_fn.__name__, partial(override_fn, original_fn))
+
+        # Don't actually modify the function being decorated.
+        return override_fn
+    return wrap
+
+
+def makeErrorFn(original_fn, suppress_categories, suppress_message_matches):
+    """ Create a return a wrapped version of the error-report function which suppresses specific
```
Comment (skallaher) [inline tools/linters/cpplint_wrapper.py]: Should probably say "Create **and** return a" not "Create a return a"

File: tools/linters/cpplint_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        # Line does not contain both an opening and closing brace.
+        m = Match(r'^(.*){(.*)$', line)
+        if m and not (IsBlankLine(m.group(1))):
+            # Line contains a starting brace and is not empty, uh oh.
+            if "=" in line:
+                # Opening brace is permissable in case of an initializer.
+                pass
+            else:
+                error(filename, linenum, 'whitespace/braces', 4,
+                      'when starting a new scope, { should be on a line by itself')
```
Comment (skallaher) [inline tools/linters/cpplint_wrapper.py]: Perhaps escaping out some single quotes to go around the brace, i.e. `'{'` just to make it obvious this is a character. Just sort of a visual thing and not necessary.

File: tools/linters/cpplint_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+def CheckIncludeLine(fn, filename, clean_lines, linenum, include_state, error):
+    """ Run the function to get include state, but suppress all the errors, since
+        ROS C++ Style is silent on include order, and contains no prohibition on use of streams. """
+    fn(filename, clean_lines, linenum, include_state,
+       makeErrorFn(error, ['build/include_order', 'build/include_alpha', 'readability/streams'], []))
+
+
+@patch(cpplint)
+def CheckSpacing(fn, filename, clean_lines, linenum, nesting_state, error):
+    """ Do most of the original Spacing checks, but suppress the ones related to braces, since
```
Comment (skallaher) [inline tools/linters/cpplint_wrapper.py]: Is there a reason that `Spacing` is capitalized?

File: CMakeLists.txt [hunk 5]
```diff
@@ -39,12 +39,20 @@ find_package(OpenCV REQUIRED)
...
-## C++:
-set(ROSLINT_CPP_OPTS "--filter=-legal,-whitespace/comments,-whitespace/parens,-build/namespaces,-readability/namespace,-runtime/threadsafe_fn" "--linelength=80")
-roslint_cpp()
+## C++ Using the custom linter to avoid the ROS overloads:
+
+# Recursively select c++ source files
+file (GLOB_RECURSE FILES *.cpp *.h *.hpp) #Set the files to parse
+
+# Run the custom linter
+# To change the options for the CPP linter edit the CPPLINT.cfg file int the
```
Comment (skallaher) [inline CMakeLists.txt:48]: Should say "in", not "int"

File: CMakeLists.txt [hunk 6]
```diff
@@ -39,12 +39,20 @@ find_package(OpenCV REQUIRED)
...
+#     src directory
+roslint_custom(${CMAKE_CURRENT_LIST_DIR}/tools/linters/cpplint "--extensions=cpp,hpp,h" ${FILES})
 
 ## Python
-set(ROSLINT_PYTHON_OPTS "--ignore=E3,E121,E123,E126,E226,E24,E704" "--max-line-length=80")
+# To change the options for the python linter edit the PYLINT.cfg file in the
+#     tools/linters directory
```
Comment (skallaher) [inline CMakeLists.txt:54]: Why is the CPPLINT.cfg file in the src directory and the PYLINT.cfg file in the linters directory? For consistency, probably either move the CPPLINT.cfg file to the linters directory or the PYLINT.cfg file to the scripts directory.

**Review by skallaher — COMMENTED** [review]
I have commented on your comments mostly and one question about file locations.

File: tools/linters/cpplint_wrapper.py (hunk 4 above)
Comment (bkallaher) [reply tools/linters/cpplint_wrapper.py]: Not my comment

File: tools/linters/cpplint_wrapper.py (hunk 3 above)
Comment (bkallaher) [reply tools/linters/cpplint_wrapper.py]: Not my comment

File: tools/linters/cpplint_wrapper.py (hunk 2 above)
Comment (bkallaher) [reply tools/linters/cpplint_wrapper.py]: Not my comment

File: tools/linters/cpplint_wrapper.py (hunk 1 above)
Comment (bkallaher) [reply tools/linters/cpplint_wrapper.py]: Not my comment

Comment (skallaher) [reply tools/linters/cpplint_wrapper.py]: These may not be your comments, but it shouldn't be too hard to fix.

File: CMakeLists.txt (hunk 6 above)
Comment (bkallaher) [reply CMakeLists.txt:54]: The fewer files in the src directory the better, I have the ability to choose the location for the python configuration but not the cpp configuration as a cpp configuration can be recursively defined by adding a configuration in any source directory in the tree. This allows for different directories to have different settings.

File: tools/linters/cpplint_wrapper.py (hunk 1 above)
Comment (bkallaher) [reply tools/linters/cpplint_wrapper.py]: There is no reason to fix them as no one should be modifying the files past this point.

File: tools/linters/CPPLINT.cfg [hunk 7]
```diff
@@ -0,0 +1,3 @@
+set noparent
```
Comment (irwineffect) [inline tools/linters/CPPLINT.cfg]: Is this file necessary? Can we just have one?

**Review by irwineffect — CHANGES_REQUESTED** [review]

**Review by irwineffect — APPROVED** [review]
